File: nestjs/07-nest-continent-crud-validation/src/modules/continent/continent.service.ts
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { ContinentRepository } from './continent.repository';

@Injectable()
export class ContinentService {
  constructor(private readonly continentRepository: ContinentRepository) {}

  getAll() {
    return this.continentRepository.findAll();
  }

  getById(id: number) {
    const continent = this.continentRepository.findById(id);
    if (!continent) {
      throw new NotFoundException(`Continent ${id} not found`);
    }
    return continent;
  }

  create(data: any) {
    if (!data?.name) {
      throw new BadRequestException('name is required');
    }
    return this.continentRepository.create({ name: data.name });
  }

  update(id: number, data: any) {
    const updated = this.continentRepository.update(id, data);
    if (!updated) {
      throw new NotFoundException(`Continent ${id} not found`);
    }
    return updated;
  }

  delete(id: number) {
    const deleted = this.continentRepository.delete(id);
    if (!deleted) {
      throw new NotFoundException(`Continent ${id} not found`);
    }
    return true;
  }
}

File: springboot-java-25/07-springboot-repository-jpa/src/main/java/com/ganatan/starter/continents/ContinentService.java
package com.ganatan.starter.continents;

import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class ContinentService {

  private final ContinentRepository repository;

  public ContinentService(ContinentRepository repository) {
    this.repository = repository;
  }

  public List<Continent> getAll() {
    return repository.findAll();
  }

  public Continent getById(int id) {
    return repository.findById(id)
      .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "CONTINENT_NOT_FOUND"));
  }

  public Continent create(String name) {
    return repository.save(name);
  }

  public Continent update(int id, String name) {
    return repository.update(id, name)
      .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "CONTINENT_NOT_FOUND"));
  }

  public void delete(int id) {
    if (!repository.deleteById(id)) {
      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "CONTINENT_NOT_FOUND");
    }
  }
}
